src/toplevel.c

```diff
@@ -339,99 +336,87 @@ static int jl_eval_expr_with_compiler_p(jl_value_t *e, int compileloops, jl_modu
         m = from;
         while (1) {
             if (i >= jl_array_len(args))
                 jl_error("invalid module path");
             var = (jl_sym_t*)jl_array_ptr_ref(args, i);
-            if (!jl_is_symbol(var))
-                jl_type_error("import or using", (jl_value_t*)jl_sym_type, (jl_value_t*)var);
+            if (var != dot_sym)
+                break;
             i++;
-            if (var != dot_sym) {
-                if (i == jl_array_len(args))
-                    return m;
-                else
-                    break;
-            }
             m = m->parent;
         }
     }
 
     while (1) {
-        if (jl_binding_resolved_p(m, var)) {
-            jl_binding_t *mb = jl_get_binding(m, var);
-            jl_module_t *m0 = m;
-            int isimp = jl_is_imported(m, var);
-            assert(mb != NULL);
-            if (mb->owner == m0 || isimp) {
-                m = (jl_module_t*)mb->value;
-                if ((mb->owner == m0 && m != NULL && !jl_is_module(m)) ||
-                    (isimp && (m == NULL || !jl_is_module(m))))
-                    jl_errorf("invalid module path (%s does not name a module)",
-                              jl_symbol_name(var));
-                // If the binding has been resolved but is (1) undefined, and (2) owned
-                // by the module we're importing into, then allow the import into the
-                // undefined variable (by setting m back to m0).
-                if (m == NULL)
-                    m = m0;
-                else
-                    break;
-            }
-        }
-        if (m == jl_main_module) {
-            if (!retrying && i==1) { // (i==1) => no require() for relative imports
-                if (require_func == NULL && jl_base_module != NULL)
-                    require_func = jl_get_global(jl_base_module, jl_symbol("require"));
-                if (require_func != NULL) {
-                    jl_value_t *reqargs[2] = {require_func, (jl_value_t*)var};
-                    jl_apply(reqargs, 2);
-                    return eval_import_path_(from, args, 1);
-                }
-            }
-        }
-        if (retrying && require_func) {
-            jl_printf(JL_STDERR, "WARNING: requiring \"%s\" in module \"%s\" did not define a corresponding module.\n",
-                      jl_symbol_name(var),
-                      jl_symbol_name(from->name));
-            return NULL;
-        }
-        else {
-            jl_errorf("in module path: %s not defined", jl_symbol_name(var));
-        }
-    }
-
-    for(; i < jl_array_len(args)-1; i++) {
-        jl_value_t *s = jl_array_ptr_ref(args,i);
-        assert(jl_is_symbol(s));
-        m = (jl_module_t*)jl_eval_global_var(m, (jl_sym_t*)s);
+        var = (jl_sym_t*)jl_array_ptr_ref(args, i);
+        if (!jl_is_symbol(var))
+            jl_type_error("import or using", (jl_value_t*)jl_sym_type, (jl_value_t*)var);
```

Comment:
pass the right name (import or using) as an argument? not super important, but seems like it should be trivial enough